### ISSUE TITLE
Bug in sox that effects OSX native_client

### ISSUE THREAD

**Issue by kdavis-mozilla** [mozilla/DeepSpeech#461]
Use of the native client on OS X exposes a sox bug that we need to work around. 

TL;DR: `fmemopen` and `open_memstream` don't exist on OS X. So we must require as a prerequisite a replacement, something like [memorymapping](https://github.com/NimbusKit/memorymapping), a `fmemopen` port, on OS X.

The long version...

When running the client on OS X one finds
```bash
(tensorflow) kdaviss-MBP:native_client kdavis$ ARGS="/Users/kdavis/Code/temp/DeepSpeech/export_dir/output_graph.pb /Users/kdavis/Code/temp/DeepSpeech/data/ldc93s1/LDC93S1.wav" make run
DYLD_LIBRARY_PATH=../../tensorflow/bazel-bin/tensorflow:../../tensorflow/bazel-bin/native_client: ./deepspeech /Users/kdavis/Code/temp/DeepSpeech/export_dir/output_graph.pb /Users/kdavis/Code/temp/DeepSpeech/data/ldc93s1/LDC93S1.wav
2017-03-24 12:41:03.668250: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-03-24 12:41:03.668838: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-03-24 12:41:03.668843: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-03-24 12:41:03.668847: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-03-24 12:41:03.668851: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
formats: can't open output file `': No such file or directory
Assertion failed: (output), function main, file client.cc, line 102.
make: *** [run] Abort trap: 6
```
This occurs as the client's call to `sox_open_memstream_write()` fails.

In particular the native client calls `sox_open_memstream_write()` in [client.cc#L98](https://github.com/mozilla/DeepSpeech/blob/master/native_client/client.cc#L98) as follows:
```C
...
  char *buffer;
  size_t buffer_size;
  sox_format_t* output = sox_open_memstream_write(&buffer, &buffer_size,
                                                  &target_signal,
                                                  &target_encoding,
                                                  "raw", NULL);
  assert(output);
...
```
In turn `sox_open_memstream_write()` is implemented as follows [formats.c#L961](https://github.com/chirlu/sox/blob/4927023d0978615c74e8a511ce981cf4c29031f1/src/formats.c#L961)
```C
sox_format_t * sox_open_memstream_write(
    char                     * * buffer_ptr,
    size_t                   * buffer_size_ptr,
    sox_signalinfo_t   const * signal,
    sox_encodinginfo_t const * encoding,
    char               const * filetype,
    sox_oob_t          const * oob)
{
  return open_write("", NULL, (size_t)0, buffer_ptr, buffer_size_ptr, signal, encoding, filetype, oob, NULL);
}
```
Then `open_write()` is implemented as follows [formats.c#L819](https://github.com/chirlu/sox/blob/4927023d0978615c74e8a511ce981cf4c29031f1/src/formats.c#L819)
```C
static sox_format_t * open_write(
    char               const * path,
    void                     * buffer UNUSED,
    size_t                     buffer_size UNUSED,
    char                     * * buffer_ptr UNUSED,
    size_t                   * buffer_size_ptr UNUSED,
    sox_signalinfo_t   const * signal,
    sox_encodinginfo_t const * encoding,
    char               const * filetype,
    sox_oob_t          const * oob,
    sox_bool           (*overwrite_permitted)(const char *filename))
{
 ...
      ft->fp =
#ifdef HAVE_FMEMOPEN
        buffer? fmemopen(buffer, buffer_size, "w+b") :
        buffer_ptr? open_memstream(buffer_ptr, buffer_size_ptr) :
#endif
        fopen(path, "w+b");
      if (ft->fp == NULL) {
        lsx_fail("can't open output file `%s': %s", path, strerror(errno));
        goto error;
      }
    }
```
But `fmemopen()` and `open_memstream()` do not exist on OS X. So this fails as `ft->fp` is `NULL` due to the fact that `path` is the empty string `""`.

My guess is that we must require as a prerequisite something like [memorymapping](https://github.com/NimbusKit/memorymapping), a `fmemopen` port, on OS X.

**Comment by kdavis-mozilla** [comment]
@Cwiiis How do you think we should handle this? My thoughts were to:

- Use `sox_open_write()` and not `sox_open_memstream_write()`
- Write a `fmemopen()` and `open_memstream()` replacement, require it as an OS X prerequisite
- Lobby sox to fix their code, a problem they've know about for at least 6 years[[1](https://sourceforge.net/p/sox/mailman/message/27569554/)]
- Lobby Apple to adhere to POSIX 2008[[2](http://pubs.opengroup.org/onlinepubs/9699919799/functions/fmemopen.html)][[3](http://pubs.opengroup.org/onlinepubs/9699919799/functions/open_memstream.html)] 😉

My guess is that the easy way out is to use `sox_open_write()` and not `sox_open_memstream_write()`.

**Comment by Cwiiis** [comment]
Ugh... I think the best solution here, I guess, would be to calculate the size of the transcoded data and use `sox_open_mem_write`. We should be able to know this using the values present in the `sox_signal_info_t input->signal`. A bit disappointing to have to do, but not the worst... I'll write a patch. I'll leave [2] and [3] down to you in the meantime :)

**Comment by Cwiiis** [comment]
@kdavis-mozilla when you have time, could you try https://github.com/Cwiiis/DeepSpeech/tree/issue461-mac-sucks please?

**Comment by Cwiiis** [comment]
Doh, as we discussed, this has the same problem. I'm thinking we can fix this by having sox output to stdout, redirecting stdout to a pipe and reading that back into a buffer... That sucks, so I think I'll make that path Mac-only.

**Comment by Cwiiis** [comment]
Fixed in #468
